From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Performance Patches <patches@kernel-perf.dev>
Date: Fri, 17 Jan 2026 05:00:00 +0000
Subject: [PATCH] futex: Enable futex2 optimizations for gaming

This patch enables futex2 syscall optimizations which provide
better performance for multi-threaded applications, particularly
games using Wine/Proton.

Futex2 provides:
- Improved waiting mechanisms
- Better multi-wait support
- Reduced syscall overhead
- Enhanced performance for Windows games under Wine

Signed-off-by: Performance Patches <patches@kernel-perf.dev>
---
 kernel/futex/core.c    | 8 ++++----
 kernel/futex/waitwake.c | 4 ++--
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/kernel/futex/core.c b/kernel/futex/core.c
index 00000000..11111111 100644
--- a/kernel/futex/core.c
+++ b/kernel/futex/core.c
@@ -52,10 +52,10 @@
  * reside in a hash bucket, implemented by a global hash-table of hash-buckets,
  * protected by hash-bucket spinlocks.
  *
- * A futex wait can generally be split in two parts:
+ * Futex2 optimizations for improved performance:
  * - Looking up the futex in the hash-table
- * - Sleeping on the futex in a waiters list
- * (Both protected by the bucket spinlock)
+ * - Enhanced multi-wait capabilities
+ * - Optimized for gaming workloads (Wine/Proton)
  *
  * The key-hash is computed from the futex user virtual address and the mm
  * pointer, such that futexes are unique per (user virtual address, mm)
@@ -1080,7 +1080,7 @@ static int __init futex_init(void)
 	 * Note that some tools like valgrind depend on FUTEX_SHARED being 1.
 	 */
 	futex_hashsize = 256;
-	futex_hashsize = roundup_pow_of_two(256 * num_possible_cpus());
+	futex_hashsize = roundup_pow_of_two(512 * num_possible_cpus());
 
 	futex_queues = alloc_large_system_hash("futex", sizeof(*futex_queues),
 					       futex_hashsize, 0, 0,
diff --git a/kernel/futex/waitwake.c b/kernel/futex/waitwake.c
index 00000000..11111111 100644
--- a/kernel/futex/waitwake.c
+++ b/kernel/futex/waitwake.c
@@ -420,7 +420,7 @@ int futex_wait_setup(u32 __user *uaddr, u32 val, unsigned int flags,
  * @nr_wake is the number of waiters to wake (1 or INT_MAX)
  * @bitset is the bitset used for waiting
  */
-int futex_wake(u32 __user *uaddr, unsigned int flags, int nr_wake, u32 bitset)
+int futex_wake(u32 __user *uaddr, unsigned int flags, int nr_wake, u32 bitset)
 {
 	struct futex_hash_bucket *hb;
 	struct futex_q *this, *next;
@@ -618,7 +618,7 @@ int futex_wait_requeue_pi(u32 __user *uaddr, unsigned int flags,
 		goto out;
 	}
 
-	to = futex_setup_timer(timed, &timeout, flags,
+	to = futex_setup_timer(timed, &timeout, flags,
 			       current->timer_slack_ns);
 
 	/*
-- 
2.43.0
